plugin/storage/es/esRollover.py

```diff
@@ -57,33 +60,36 @@ def main():
     if str2bool(os.getenv('ARCHIVE', 'false')):
         write_alias = prefix + ARCHIVE_INDEX + '-write'
         read_alias = prefix + ARCHIVE_INDEX + '-read'
-        perform_action(action, client, write_alias, read_alias, prefix+'jaeger-span-archive', 'jaeger-span')
+        perform_action(action, client, write_alias, read_alias, prefix+'jaeger-span-archive', 'jaeger-span', prefix)
     else:
         write_alias = prefix + 'jaeger-span-write'
         read_alias = prefix + 'jaeger-span-read'
-        perform_action(action, client, write_alias, read_alias, prefix+'jaeger-span', 'jaeger-span')
+        perform_action(action, client, write_alias, read_alias, prefix+'jaeger-span', 'jaeger-span', prefix)
         write_alias = prefix + 'jaeger-service-write'
         read_alias = prefix + 'jaeger-service-read'
-        perform_action(action, client, write_alias, read_alias, prefix+'jaeger-service', 'jaeger-service')
+        perform_action(action, client, write_alias, read_alias, prefix+'jaeger-service', 'jaeger-service', prefix)
 
 
-def perform_action(action, client, write_alias, read_alias, index_to_rollover, template_name):
+def perform_action(action, client, write_alias, read_alias, index_to_rollover, template_name, prefix):
     if action == 'init':
         shards = os.getenv('SHARDS', SHARDS)
         replicas = os.getenv('REPLICAS', REPLICAS)
         esVersion = get_version(client)
+        use_ilm = str2bool(os.getenv("ES_USE_ILM", 'false'))
         if esVersion == 7:
             mapping = Path('./mappings/'+template_name+'-7.json').read_text()
         else:
             mapping = Path('./mappings/'+template_name+'.json').read_text()
-        create_index_template(fix_mapping(mapping, shards, replicas), template_name)
+            use_ilm = False
+        create_index_template(fix_mapping(mapping, shards, replicas, prefix, use_ilm),
+                              prefix+template_name+"-override-template")
```

Comment:
Why is the `-override-template` necessary here? It seems to add another template matching the same index pattern; but the only difference is the `order`; which seems unnecessary to me.

Reply:
@albertteoh  - Kept this because, as of now if some one wants to override replicas/shards - the only way they could do is with esRollover script (passing different Shard and Replica counts). By default, I guess counts are hard coded.

https://github.com/jaegertracing/jaeger/blob/0b43839844eb478f2b0d328495944a2e16f9a519/plugin/storage/es/factory.go#L216

https://github.com/jaegertracing/jaeger/blob/0b43839844eb478f2b0d328495944a2e16f9a519/plugin/storage/es/factory.go#L217

Reply:
If we dont have _-override-template_  suffixed to the index template name. When jaeger comes up, if --es.create-index-templates is not set to false, it would overwrite the template created by rollover script (which is always run before jaeger is brought up) - as the name of index template would be same.

Reply:
@bhiravabhatla
> Kept this because, as of now if some one wants to override replicas/shards - the only way they could do is with esRollover script (passing different Shard and Replica counts).

Yup, this is a use case which I hadn't considered. However, if we do want to give users ability to override the shard and replica counts (presumably via the jaeger-collector), shouldn't `Order` be set to `0`? From my own tests, because order for `-override-template` is `2`, my jaeger-collector shard and replica settings (which have `order = 1`) don't take effect because they are overridden by the esRollover script's `-override-template`. Recall the templates are merged/applied in the order: 0 -> 1 -> 2 -> .... They are additively applied and the larger `order` is preferred on conflict/overlapping configuration.

> If we dont have -override-template suffixed to the index template name. When jaeger comes up, if --es.create-index-templates is not set to false, it would overwrite the template created by rollover script

Yes, you're right. Is this necessarily a bad thing though? From what I understand, the trade-offs of overwriting the template created by esRollover script are:
**Pros**:
- Reduces the number of overlapping templates:
  - Simplifies the development & debugging process; there isn't a need to mentally merge two similar templates nor a need to understand how ordering works with ES templates.
  - Avoids potential bugs from overlapping json elements that were not considered.
- Removes the need to manage ordering, which in turn, can avoid potential bugs if users decide to create their own templates matching the same jaeger indices, but with the wrong order. This also simplifies the code, removing the need to manage template `order`ing.
- Provides the ability to easily "off-board" from ILM, if users wish to customise management of index rollovers because templates are _additively_ applied when overlapping with another template.
  - Consider the scenario where a user creates indices against an ILM-enabled template. They then decide to revert to manual rollovers (perhaps to have finer-grain control on index management). However, they cannot disable ILM, even when setting `ES_USE_ILM=false` in jaeger-collector, because the base template contains the `lifecycle` setting and can't be removed through overriding templates.

**Cons**:
- Overwriting means we will lose history of the original template and when it was replaced.
- The base template may have index configurations that users may want and overwriting will remove these. 
  - Addressable by installing the custom template and running jaeger-collector with `--es.create-index-templates`, as suggested in: #1077.

To me, the benefits of simplicity, maintainability and the ability of transition from ILM to manual rollovers outweigh the costs, though perhaps my understanding is incorrect. Let me know what you think.

Reply:
@albertteoh  
> However, if we do want to give users ability to override the shard and replica counts (presumably via the jaeger-collector)

As of now there is no way to override  the shard and replica counts  from collector. Its always hardcoded to 10. Only way to achieve it is via override template in esRollover. And the ordering works as you explained. So, only way to customize shards/replicas as of now is to give higher order to templates getting created from esRollover or have additional flags in collector to customize the counts. Not sure if I am missing something

Reply:
And I agree with your points on ease of maintenance/on-boarding and off-boarding. Also I see esRollover script is trying to solve multiple problems. If you agree, we can have a separate issue created to add support for customizing replicas and shards from collector flags and then we can jus use esRollover for manual rollover and creation of initial aliases/indices

Reply:
> As of now there is no way to override the shard and replica counts from collector.

I can see it's possible and was able to override shard counts in my template via the collector:

```
$ SPAN_STORAGE_TYPE=elasticsearch ES_NUM_SHARDS=3 go run ./cmd/collector/main.go
```

Help docs:
```
$ SPAN_STORAGE_TYPE=elasticsearch go run ./cmd/collector/main.go --help
...
      --es.num-replicas int                                The number of replicas per index in Elasticsearch (default 1)
      --es.num-shards int                                  The number of shards per index in Elasticsearch (default 5)
...
```

Reply:
Oh, I missed this then. My bad.

I see it now - 

```golang
spanMapping, serviceMapping := GetSpanServiceMappings(cfg.GetNumShards(), cfg.GetNumReplicas(), client.GetVersion(), cfg.GetIndexPrefix())
```

Reply:
Will push the changes removing the suffix in template name and order in the template.

Reply:
Implemented above feedback @albertteoh